446 improve res mod perf

hyi opened the pull request:
This branch is slated for NFIE hotfix release, and deployed on park for testing. Please review the code at your earliest convenience so that it can be merged into develop for the NFIE hotfix. It mainly resolves the performance issue with big resource metadata editing by doing on-demand bagging. Another two issues below are also resolved in this branch:

1. Hide the ajax spin wheel icon when irods browser gets invoked again - a minor bug fix.
2. bagit rule fix that stores generated bagit files such as bagit.txt and manifest files to HydroShare default resource rather than the default iRODS demoResc resource.

selimnairb commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -21,20 +22,14 @@ def delete_bag(resource):
+    # delete the bags table
+    for bag in resource.bags.all():
+        try:
+            bag.delete()
```
remove whitespace between *try:* and *bag.delete()*.

selimnairb commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -21,20 +22,14 @@ def delete_bag(resource):
+        try:
+            bag.delete()
+        except Exception as ex:
+            raise Exception(ex.message)
```
Remove whitespace before *raise ...*

pkdash replied:
Is the use of ```try ... except``` here to express the intent that bag.delete() may fail?

ironfroggy replied:
This `try/except` does nothing but obscure the traceback.

hyi replied:
@pkdash @ironfroggy The previous code used try/except for bag.delete, but rather than raise an exception, it just ignores/suppresses the error by using "pass" in except: block. I thought it is better to raise the error, but now that I think about it more, the error is raised anyway, so I should not add another layer. So I have removed try/except block here and pushed the code. If any of you see we should ignore the exception here as we did before, let me know as I think we should raise the exception here rather than ignore it.

selimnairb commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -12,6 +12,7 @@
 
 def delete_bag(resource):
+
     res_id = resource.short_id
```
remove whitespace before *res_id...*

selimnairb commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -21,20 +22,14 @@ def delete_bag(resource):
 
-    :return: the hs_core.models.Bags instance associated with the new bag.
-    """
+def create_bag_files(resource):
```
need docstrings documenting this function.

selimnairb commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -130,16 +125,43 @@ def create_bag(resource):
     to_file_name = os.path.join(resource.short_id, 'data', 'resourcemap.xml')
     istorage.saveFile(from_file_name, to_file_name, False)
```
How do we know whether istorage.saveFile() succeeded or failed?

hyi replied:
saveFile() runs an irods iput session which will raise SessionException if it fails.

selimnairb commented on hydroshare/irods/ruleGenerateBagIt_HS.r:
```diff
@@ -74,5 +74,5 @@ generateBagIt {
   msiWriteRodsLog("BagIt bag files created in place: *BAGITDATA <- *BAGITDATA",*Status);
 }
-INPUT *BAGITDATA="/dummy/dummy/dummy" 
+INPUT *BAGITDATA="/hydrotest41Zone/home/hyi", *DESTRESC="demoResc"
```
This looks like it is specific to your development environment.

hyi replied:
This is just a dummy placeholder, and will be overriden by parameters being passed in at run time, but I will change it to be "dummy" to make it clear.

selimnairb commented on hs_core/hydroshare/hs_bagit.py:
```diff
@@ -21,20 +22,14 @@ def delete_bag(resource):
+def create_bag_files(resource):
     from . import utils as hs_core_utils
     DATE_FORMAT = "YYYY-MM-DDThh:mm:ssTZD"
     istorage=IrodsStorage()
```
Regarding:

    dest_prefix = getattr(settings, 'BAGIT_TEMP_LOCATION', '/tmp/hydroshare/')

What happens if the directory represented by dest_prefix does not exist?

hyi replied:
the directory will be created if it does not exist.

hyi commented:
I have updated the code to use iRODS metadata AVU pairs to keep track of whether a bag is modified for a resource and also used iRODS metadata AVU to lock the bag while the bag is being updated to avoid potential race condition that could result from multiple requests modifying the same bag. Also changed to not create the bag initially when the resource is created, but rather to set the bag_modified AVU metadata to True for the resource so that the bag will be created when it is downloaded. Tested all scenarios including the potential race condition and all seem to work as expected. Will ask Michael to update the code on park.hydroshare.org for team to test. In the meanwhile, please review the code at your earliest convenience so that it can be merged for NFIE hotfix release.

selimnairb commented:
@hyi I tested the REST API against park.hydroshare.org using hs_restclient.  Resource creation, making public, and deletion all work as expected.

@valentinedwv Are we ready to do any load testing using hs_restclient?  This would be useful to test for race conditions in the resource bag AVU locking and spin lock code that @hyi wrote.

hyi commented:
Have implemented everything I can do for the on-demand bagging for the NFIE hotfix (which most likely needs to be deployed this coming Monday) and ask for final code review. The latest changes I made include:
1. used irods metadata AVU to keep track of whether a resource is dirty with its bag to be re-created when downloading.
2. removed AJAX download to revert it back to the previous url/view mapping for download, hence spin wheel progress indicator does not show any more.
3. added more check and exception handling for potential race conditions that could result from clashes between multiple irods processes triggered from multiple requests, e.g., when user A is downloading a resource bag that triggers bag-creation on iRODS side, user B deletes this same resource or modifies metadata of this same resource, and initiates a download also at the same time. This is only workaround implemented to avoid exceptions raised due to these potential race conditions. In the future, we need to use external services/packages to ensure process synchronization as recommended by iRODS expert Hao Xu.

hyi commented:
The latest code update has been deployed on park.hydroshare.org by Michael for team testing.

dtarb commented:
@hyi thanks for your work on this.  Given that you posted this on Friday, lets not rush a Monday deploy.  Lets review it properly and deploy after review and testing.

selimnairb commented on hs_core/templates/pages/genericresource.html:
```diff
@@ -1416,6 +1416,9 @@ <h4 class="modal-title">Confirm file deletion</h4>
             </div>
             </div>
         <!-- ==== Tabs end ==== -->
+        <div class="rating">
```
how is this related to the hotfix?

hyi replied:
This is another fix I slated in for a regression error that was probably introduced by some conflict in merging along the way (this issue was raised in github by @Castronova). Specifically, the rating for resource code was removed (which was present on beta, but not on www), so I am adding the code back so that users can rate resources as well as rate comments to resolve that regression error. My branch commit history should give more details on this.

selimnairb commented on irods_browser_app/static/js/irods.js:
```diff
@@ -34,6 +34,7 @@ $('#btn-signin-irods').on('click',function(){
 $('#btn-select-irods-file').on('click',function(){
     $('#res_type').val($('#resource-type').val());
     $('#file_struct').children().remove();
+    $('.ajax-loader').hide();
```
This appears to be leftover from when you were using the AJAX loader.

hyi replied:
@selimnairb No, this is not a left-over code, but instead it is a minor fix I slated in for iRODS browser I put in which also shows the same spin-wheel when retrieving collection and data objects from iRODS for browsing purposes. I found I did not hide it properly which will have spin-wheel being displayed from start to finish if the user click browse button again. This one line addition will fix that issue.

selimnairb commented:
Pending @ironfroggy's review, I think this is ready to merge.  (though this is not the last word on the subject of downloads from iRODS, though I think @hyi has done a good job to improve metadata editing performance.)
